src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs

```diff
@@ -395,46 +395,38 @@ public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> source, out Rune
             // it tries to consume as many code units as possible as long as those code
             // units constitute the beginning of a longer well-formed subsequence per Table 3-7.
 
-            int index = 0;
-
-            // Try reading input[0].
+            // Try reading source[0].
 
-            if ((uint)index >= (uint)source.Length)
+            int index = 0;
+            if (source.IsEmpty)
             {
                 goto NeedsMoreData;
             }
 
-            uint tempValue = source[index];
-            if (!UnicodeUtility.IsAsciiCodePoint(tempValue))
+            uint tempValue = source[0];
+            if (UnicodeUtility.IsAsciiCodePoint(tempValue))
             {
-                goto NotAscii;
+                bytesConsumed = 1;
+                result = UnsafeCreate(tempValue);
+                return OperationStatus.Done;
             }
 
-        Finish:
-
-            bytesConsumed = index + 1;
-            Debug.Assert(1 <= bytesConsumed && bytesConsumed <= 4); // Valid subsequences are always length [1..4]
-            result = UnsafeCreate(tempValue);
-            return OperationStatus.Done;
-
-        NotAscii:
-
             // Per Table 3-7, the beginning of a multibyte sequence must be a code unit in
             // the range [C2..F4]. If it's outside of that range, it's either a standalone
             // continuation byte, or it's an overlong two-byte sequence, or it's an out-of-range
             // four-byte sequence.
 
+            // Try reading source[1].
+
+            index = 1;
             if (!UnicodeUtility.IsInRangeInclusive(tempValue, 0xC2, 0xF4))
             {
-                goto FirstByteInvalid;
+                goto Invalid;
             }
 
             tempValue = (tempValue - 0xC2) << 6;
 
-            // Try reading input[1].
-
-            index++;
-            if ((uint)index >= (uint)source.Length)
+            if (1 >= (uint)source.Length)
```

Comment:
With recent JIT-improvements this can be written less yoda-like.
```suggestion
            if (source.Length <= 1)
```
The same on other places (<= 2 L483, <= 3 L508).

Cf. https://github.com/dotnet/runtime/issues/67044#issuecomment-1076509592
PR https://github.com/dotnet/runtime/pull/67448 does this for the rest of that file.